### ISSUE TITLE
v4.1.x: Failure on btl tcp on the greq1 benchmark

### ISSUE THREAD

**Issue by alinask** [open-mpi/ompi#8340]
### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)

```
$ompi_info -V
Open MPI v4.1.1a1
```

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

From a git clone.

### Please describe the system on which you are running

* Operating system/version: Red Hat Enterprise Linux Server release 7.6 (Maipo)
* Computer hardware: Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz
* Network type: IB

-----------------------------

## Details of the problem

There is a failure that happens on 2 ranks with `-mca pml ob1 -mca btl self,tcp` with an mpich 3.4 benchmark:
Reproduces on every launch with the following command line.

```
$mpirun -np 2 --display-map --map-by node -mca coll ^hcoll -mca pml ob1 -mca btl self,tcp -mca osc ^ucx  ./mpich-3.4/test/mpi/pt2pt/greq1
 Data for JOB [1320,1] offset 0 Total slots allocated 56

 ========================   JOB MAP   ========================

 Data for node: jazz14  Num slots: 28   Max slots: 0    Num procs: 1
        Process OMPI jobid: [1320,1] App: 0 Process rank: 0 Bound: N/A

 Data for node: jazz15  Num slots: 28   Max slots: 0    Num procs: 1
        Process OMPI jobid: [1320,1] App: 0 Process rank: 1 Bound: N/A

 =============================================================
Generalized request marked as complete
[jazz15:135231] *** Process received signal ***
[jazz15:135231] Signal: Segmentation fault (11)
[jazz15:135231] Signal code: Address not mapped (1)
[jazz15:135231] Failing at address: 0x1
[jazz15:135231] [ 0] /lib64/libpthread.so.0(+0xf5d0)[0x7f2a7895a5d0]
[jazz15:135231] [ 1] /hpc/local/benchmarks/daily/next/2021-01-04/hpcx-gcc-redhat7.6/ompi/lib/libmpi.so.40(ompi_grequest_complete+0xde)[0x7f2a78bb324e]
[jazz15:135231] [ 2] Generalized request marked as complete
[jazz14:194599] *** Process received signal ***
[jazz14:194599] Signal: Segmentation fault (11)
[jazz14:194599] Signal code: Address not mapped (1)
[jazz14:194599] Failing at address: 0x1
/hpc/local/benchmarks/daily/next/2021-01-04/hpcx-gcc-redhat7.6/ompi/lib/libmpi.so.40(MPI_Grequest_complete+0x9)[0x7f2a78bd0f39]
[jazz15:135231] [ 3] /labhome/alinas/workspace/MTT-repo/hpc-mtt-conf/mpich-3.4/test/mpi/pt2pt/greq1[0x401da1]
[jazz15:135231] [ 4] /lib64/libc.so.6(__libc_start_main+0xf5)[0x7f2a785a03d5]
[jazz15:135231] [ 5] /labhome/alinas/workspace/MTT-repo/hpc-mtt-conf/mpich-3.4/test/mpi/pt2pt/greq1[0x401e71]
[jazz15:135231] *** End of error message ***
[jazz14:194599] [ 0] /lib64/libpthread.so.0(+0xf5d0)[0x7f11272395d0]
[jazz14:194599] [ 1] /hpc/local/benchmarks/daily/next/2021-01-04/hpcx-gcc-redhat7.6/ompi/lib/libmpi.so.40(ompi_grequest_complete+0xde)[0x7f112749224e]
[jazz14:194599] [ 2] /hpc/local/benchmarks/daily/next/2021-01-04/hpcx-gcc-redhat7.6/ompi/lib/libmpi.so.40(MPI_Grequest_complete+0x9)[0x7f11274aff39]
[jazz14:194599] [ 3] /labhome/alinas/workspace/MTT-repo/hpc-mtt-conf/mpich-3.4/test/mpi/pt2pt/greq1[0x401da1]
[jazz14:194599] [ 4] /lib64/libc.so.6(__libc_start_main+0xf5)[0x7f1126e7f3d5]
[jazz14:194599] [ 5] /labhome/alinas/workspace/MTT-repo/hpc-mtt-conf/mpich-3.4/test/mpi/pt2pt/greq1[0x401e71]
[jazz14:194599] *** End of error message ***
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun noticed that process rank 1 with PID 135231 on node jazz15 exited on signal 11 (Segmentation fault).
--------------------------------------------------------------------------
2 total processes killed (some possibly by mpirun during cleanup)
```

The benchmark can be downloaded from https://www.mpich.org/downloads/
(mpich-3.4 (stable release))
```
cd mpich-3.4
./autogen.sh
cd test/mpi
./configure  --with-mpi=<path/to/ompi/install> --enable-strictmpi --disable-spawn --disable-cxx
cd pt2pt/
make
```

**Comment by alinask** [comment]
[greq1.txt](https://github.com/open-mpi/ompi/files/5774862/greq1.txt)

**Comment by bosilca** [comment]
I don't think the reported error has anything to do with any BTL or PML, simply because we are talking about a standalone test checking the behavior of generalized requests. #8347 fixes the issue for me.

**Comment by jsquyres** [comment]
The master PR was merged; re-opening this until a v4.1.x PR is merged.

**Comment by jsquyres** [comment]
PR #8348 filed for v4.1.x.  @alinask can you review?

**Comment by alinask** [comment]
> The master PR was merged; re-opening this until a v4.1.x PR is merged.

@jsquyres As the v4.1.x PR is merged, I'm closing this ticket.
Thanks!
